RFD 0077 Support Dump - early first draft

programmerq opened the pull request:
Updates #4224

programmerq committed:
RFD 0077 Support Dump - early first draft

klizhentas reviewed (commented):
I think it's a pretty good idea.

Why not split collection and representation? I am assuming there will be a lot of different data points collected, so dump can be as simple as writing everything to a sqlite.db file. Then you can have a special program that will read the database and produce the readable output via CLI or UI.

This will allow you to split the complexity of rendering/navigating the contents of the database.

This approach is identical to `go tool pprof`

zmb3 commented on rfd/0077-support-dump.md:
```diff
@@ -0,0 +1,252 @@
+may be transmitted over systems with extremely limited file sizes.
+
+The end user should be able to run one command or click on one web gui button
+to generate a dump file. This file will likely need to be a compressed tar file.
```
Remove the note about compressed tar file. That's an implementation detail and not really related to "End User Experience."

zmb3 commented on rfd/0077-support-dump.md:
```diff
@@ -0,0 +1,252 @@
+It should be built in to the Teleport binary itself. It should be possible to
+initiate a support dump from both the command line or from the web interface.
+Provisions should be made to advise about (or redact) possible sensitive
+information where applicable.
```
We'll need to consider how this integrates with RBAC and what permissions would be needed to request a dump.

zmb3 commented on rfd/0077-support-dump.md:
```diff
@@ -0,0 +1,252 @@
+logs from standard locations wherever possible. It should advise the end user
+if it is unable to gather critical information for any reason.
+
+It should be built in to the Teleport binary itself. It should be possible to
```
I wouldn't prescribe that it has to be the `teleport` binary. `tctl` might also be a totally valid approach.

programmerq replied:
I'm of two minds on this one--

The ability to have the cluster request one from a node can be handy, but there's also times when a node won't even start. Invoking the support dump process against a stopped teleport's /var/lib/teleport could also be quite beneficial.

Thoughts?

zmb3 commented on rfd/0077-support-dump.md:
```diff
@@ -0,0 +1,252 @@
+Because teleport has many deployment methods, the support dump procedure should
+be smart enough to detect and gather information in the correct way. This may
+include an entire cluster, or a single agent. It should also be able to gather
+logs from standard locations wherever possible. It should advise the end user
```
Logs strike me as something that we'll need to explore, as they handled at a platform level and not in Teleport itself.

programmerq replied:
yes indeed-- these are often critical for troubleshooting. The support dump processes I've seen in the past will check 2-5 common places where the log may be, and then give up with a warning. This is fine as long as the warning is not silent.

I have done some experiments with trying to detect if teleport is running in systemd or in a container and then trying to see if it has access to the respective system. The mvp for this RFD will be to only check for logs immediately available to teleport. That would be places like journald or the file as configured in the teleport.yaml config.

I consider more specialized cases like getting logs from kubernetes are out of scope for this particular proposal.

zmb3 commented on rfd/0077-support-dump.md:
```diff
@@ -0,0 +1,252 @@
+
+### Contents of the Dump File
+
+* The resulting compressed tar file should have a defined directory structure.
```
Again, let's not over-prescribe implementation details. There may not be a need for directories at all.

zmb3 commented on rfd/0077-support-dump.md:
```diff
@@ -0,0 +1,252 @@
+The `objects/` directory will gather any objects it possibly can. If a support
+dump is initiated remotely with tsh credentials, it should indicate which
+objects it was unabled to gather due to permission issues. This should collect
+every object type it can possibly retrieve. The separate .error files would
```
If I understand correctly, the motivation for this feature is to help support customers who are having an issue and ultimately get to resolution faster.

Collecting anything and everything that can "possibly be retrieved" does not necessarily lead to faster time to resolution. On very large clusters, for example, you may find that this makes the dump take longer to execute, decreases performance of the cluster while it is being collected, and provides the support engineer with lots of useless information to sift through.

I would rather start by identifying the things we most commonly ask for, collecting those, and then expanding scope if we discover more things should be included.

zmb3 commented on rfd/0077-support-dump.md:
```diff
@@ -0,0 +1,252 @@
+
+#### Miscellaneous Considerations
+
+When a support dump is invoked, it should print minimal progress information.
+There should be no surprises such as timing out with a failure after five
+minutes. It should instead be obvious that the support dump is working on
+something and fail quickly if it is going to fail. Users can be told to opt in
```
Similar to above, a lot of these goals (minimal progress, failing fast, not timing out) conflict with the goal of collecting "anything that can possibly be collected"

programmerq replied:
yes indeed-- a tunable balance is the real goal. The support dump should have as much capability as is possible, but have the ability to fail fast and advise that the expected file size is large or the time to do the process will be too long. For really nasty cases, it's handy to be able to turn off the warnings and really get everything no matter how long it will take.

programmerq commented:
> Why not split collection and representation? I am assuming there will be a lot of different data points collected, so dump can be as simple as writing everything to a sqlite.db file. Then you can have a special program that will read the database and produce the readable output via CLI or UI.

Often times when working with this type of support dump, it is helpful to use regular command line tools. Having an sqlite database wrapper instead of a .tar wrapper would work, but may not be ideal for most of the uses on the technical support team.

I am certainly not against having multiple output formats. Each tool like this that I've used has always been a compressed .tar or .zip or similar, so I had that in mind from the get-go.

One of the things I would do with support dumps for a previous product I worked with was to simply drop all the contents of the .tar into a splunk free instance and watch it index the logs. I just had to untar the logs in the correct place where I had splunk was running and that was that. The splunk container had a `tar` binary so I didn't need any extra steps to deal with the actual dump file.

I'll revisit the pprof tools and adjust my thinking about this.

programmerq committed:
Incorporated lots of feedback

some simplifications, clarifications, restructuring, and a few other
changes went in to this pass.

russjones commented:
@programmerq What do you think about closing this? The last update is from 2022. If we want some of this functionality, let's reassess and open a new RFD.
